### PR TITLE
Add fromJson and mustFromJson funcs

### PR DESCRIPTION
**Pull request by mholt**
Resolves https://github.com/Masterminds/sprig/issues/35; I also found us needing this in Caddy. See https://github.com/caddyserver/caddy/issues/2736#issuecomment-570750244.

Incidentally, my editor made a number of good points that I agree with:

Screen Shot 2020-01-04 at 12 53 41 PM

I don't suppose there's any chance of fixing these... (v4?)

Also, JSON functions should probably documented along with the encoding functions, maybe?

### CODE REVIEW

File: defaults.go [hunk 1]
```diff
@@ -63,6 +63,19 @@ func coalesce(v ...interface{}) interface{} {
...
+// fromJson decodes JSON into a structured value, ignoring errors.
+func fromJson(v string) interface{} {
+	output, _ := mustFromJson(v)
+	return output
+}
+
+// mustFromJson decodes JSON into a structured value, returning errors.
+func mustFromJson(v string) (interface{}, error) {
+	var output interface{}
+	err := json.Unmarshal([]byte(v), &output)
```
Comment (twpayne) [inline defaults.go]: This will silently fail if `v` cannot be converted to a `[]byte`. The type conversion should be checked.

Comment (mholt) [reply defaults.go]: That's a compile time check. I believe all strings can be converted to byte slices. (This is not a type assertion.)

Comment (twpayne) [reply defaults.go]: Sorry, you're right. However, I would extend the function to also accept `[]byte` as an input.

Comment (mholt) [reply defaults.go]: Why?

Comment (twpayne) [reply defaults.go]: Data flowing between template functions can be of any type. It's entirely possible that someone will want to use this function with a `[]byte`, for example:

    {{ (readFile "foo.json" | fromJson).bar }}

where `readFile` returns a `[]byte`. Why not allow this?

Comment (mholt) [reply defaults.go]: We can. I was just wondering what the use case was.

**Comment by ggillies** [comment]
Hi!

I am really keen to see this merged and become usable. Is there anything I can do to help? Is the only thing outstanding here to change input type from `string` to `[]byte`?

**Comment by renjiexu** [comment]
+1, this feature could be very helpful

**Comment by pjotre86** [comment]
This feature would be incredibly useful. How can we get this done?

**Review by technosophos — APPROVED** [review]

**Comment by mattfarina** [comment]
@mholt In reference to your editors suggestions... that would change the public API of this library. That would mean a major version increment so they will have to wait until then. When this version of the library was being written Go modules were still in transition. Breaking APIs was a real issue at that time for some so we avoided it back then. This is just the context of why the capitalization is what it is.

**Comment by mholt** [comment]
That's understandable; thanks for the merge!

**Pull request merged**